sdk/communication/AzureCommunication/Source/Authentication/HMACAuthenticationPolicy.swift

```diff
@@ -0,0 +1,144 @@
+    }
+    
+    public func addAuthenticationHeaders(
+        url: URL,
+        httpMethod: String,
+        contents: Data) -> HTTPHeaders {
+        var headers: HTTPHeaders = [:]
+        headers[HMACAuthenticationPolicy.contentHashHeader] = contents.sha256
+        
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "E, dd MMM YYYY HH:mm:ss 'GMT'" // Is this the right date format? Try ISO8601DateFormatter
+        dateFormatter.timeZone = TimeZone(abbreviation: "UTC")
+        
+        let date = Date()
+        let utcNow = dateFormatter.string(from: date)
+        headers[HMACAuthenticationPolicy.dateHeader] = utcNow
+        headers[HMACAuthenticationPolicy.hostHeader] = url.host
+                
+        headers.merge(addSignatureHeader(
+                        url: url,
+                        httpMethod: httpMethod,
+                        date: utcNow,
+                        contentHashed: contents.sha256)) { (_, new) in new }
+        /**
+         After signature
+         HashMap@56 size=4
+         0:HashMap$Node@113 "date":"Wed, 07 Oct 2020 18:16:02 GMT"
+         1:HashMap$Node@114 "Authorization":"HMAC-SHA256 SignedHeaders=date;host;x-ms-content-sha256&Signature=KZD9UN4LsktsEX2e9cRp+LS2opjAtEVKqt+OzFCHh9o="
+         2:HashMap$Node@115 "x-ms-content-sha256":"YjVxGFu++f6tLM9YEVQVRmchZiYyxQ+8Bi3PXTJz2C4="
+         3:HashMap$Node@116 "host":"localhost"
+         */
+        return headers
+    }
+    
+    private func addSignatureHeader(
+        url: URL,
+        httpMethod: String,
+        date: String,
+        contentHashed: String) -> HTTPHeaders {
+        // Order of the headers are important here for generating correct signature
+        let signedHeaderNames = "\(HMACAuthenticationPolicy.dateHeader);\(HMACAuthenticationPolicy.hostHeader);\(HMACAuthenticationPolicy.contentHashHeader)"
+        let signedHeaderValues =  "\(date);\(url.host ?? "");\(contentHashed)" // date;urlhost;content hash
+        
+        // Add unit test for different ports
+        // 1 with port 1 without port
+        var pathAndQuery = url.path
+        if let query = url.query {
+            pathAndQuery += "?\(query)"
+        }
+
+        let stringToSign = "\(httpMethod.uppercased())\n\(pathAndQuery)\n\(signedHeaderValues)"
+        let signature = stringToSign.generateHmac(using: accessKey) // Is this right?
+        let hmacSHA256Format = "HMAC-SHA256 SignedHeaders=\(signedHeaderNames)&Signature=\(signature)"
+        return ["Authorization": hmacSHA256Format]
+    }
+}
+
+extension Data {
+    public var sha256: String {
+        if #available(iOS 13.0, *) {
+            let hashed = SHA256.hash(data: self)
+            return Data(hashed).base64EncodedString()
+        } else {
+            return self.hash(algorithm: .sha256).base64EncodedString()
+        }
+    }
+
+}
+
+extension String {
+    func generateHmac(using secret: String) -> String {
+        if #available(iOS 13.0, *) {
```

Comment:
will this 2 approaches produce the same Hmac values? if yes, can we just do `self.hmac(algorithm: .sha256, key: Data(base64Encoded: secret)!).base64EncodedString()`

Reply:
Both of them return the same thing. The reason I wanted to leave both implementations is I wanted to have a conversation about which one to use. Within the if is an implementation that uses `CryptoKit` whereas the bottom one uses `CommonCrypto` I know as swift continues to mature Apple is looking for developers to use `CryptoKit` more.